### ISSUE TITLE
Question about attention token slicing in blocks_to_share

### ISSUE THREAD

**Issue by l3965724-create** [yuyi-sd/UPA-RFAS#3]
Hello authors,

Thank you for sharing the code for this work. While reproducing the experiments in your paper, “When Robots Obey the Patch: Universal Transferable Patch Attacks on Vision-Language-Action Models,” I had a question about one part of the attention-processing logic and would like to ask for your clarification.

The relevant code snippet is:

def blocks_to_share(attn_blocks):
    # Average over heads -> extract text queries × visual keys -> normalize as “shares”
    shares = []
    S = attn_blocks[0].size(-1)   # total number of tokens S = P + T
    for A in attn_blocks:
        # A: (B, H, S, S) with S = P + T
        A = A.float().mean(dim=1)                   # (B, S, S)
        vis_start = S - P
        A_txt_vis = A[:, vis_start:, :vis_start]    # (B, P, T)
        share = A_txt_vis / (A_txt_vis.sum(dim=-2, keepdim=True) + eps)
        shares.append(share)
    return torch.stack(shares, dim=0).mean(dim=0)   # (B, P, T)

My understanding is that the goal here seems to be extracting the attention from text queries to visual keys, i.e., text queries attending to visual tokens. Based on the paper and the model input format, I assume the token order is roughly:

[visual tokens][text tokens]

If this is the case, the visual tokens should occupy the first P positions, and the text tokens should occupy the remaining T positions. Therefore, the text-query-to-visual-key attention would seem to correspond to:

A[:, P:, :P]

with shape (B, T, P).

However, the current implementation uses:

vis_start = S - P
A_txt_vis = A[:, vis_start:, :vis_start]

This seems to select the last P tokens as queries, which may include part of the visual tokens and text tokens, and the first S-P tokens as keys, which may include part of the visual tokens.

Could you please clarify whether my understanding of the token order and this attention slicing logic is correct? Is there any special reason for this implementation, such as the actual token order being different? Of course, if the token order here is [text tokens][visual tokens], then this implementation would make sense.

Thank you very much for your help, and thanks again for open-sourcing this work.

**Comment by z20-c** [comment]
I also have this question
